File: Emulator/src/main/core/game/node/entity/lock/ActionLocks.java
package core.game.node.entity.lock;

import core.game.node.Node;
import core.game.world.GameWorld;

import java.util.HashMap;
import java.util.Map;

/**
 * The type Action locks.
 */
public final class ActionLocks {

    private Lock movementLock = new Lock();

    private Lock teleportLock = new Lock();

    private Lock componentLock = new Lock();

    private Lock interactionLock = new Lock();

    private Lock equipmentLock = null;

    private Map<String, Lock> customLocks;

    /**
     * Instantiates a new Action locks.
     */
    public ActionLocks() {
        this.customLocks = new HashMap<>();
    }

    /**
     * Lock.
     */
    public void lock() {
        lock(Integer.MAX_VALUE - GameWorld.getTicks());
    }

    /**
     * Lock.
     *
     * @param ticks the ticks
     */
    public void lock(int ticks) {
        lockMovement(ticks);
        lockInteractions(ticks);
    }

    /**
     * Unlock.
     */
    public void unlock() {
        unlockMovement();
        unlockInteraction();
    }

    /**
     * Lock movement.
     *
     * @param ticks the ticks
     */
    public void lockMovement(int ticks) {
        movementLock.lock(ticks);
    }

    /**
     * Unlock movement.
     */
    public void unlockMovement() {
        movementLock.unlock();
    }

    /**
     * Is movement locked boolean.
     *
     * @return the boolean
     */
    public boolean isMovementLocked() {
        return movementLock.isLocked();
    }

    /**
     * Lock teleport.
     *
     * @param ticks the ticks
     */
    public void lockTeleport(int ticks) {
        teleportLock.lock(ticks);
    }

    /**
     * Unlock teleport.
     */
    public void unlockTeleport() {
        teleportLock.unlock();
    }

    /**
     * Is teleport locked boolean.
     *
     * @return the boolean
     */
    public boolean isTeleportLocked() {
        return teleportLock.isLocked();
    }

    /**
     * Lock component.
     *
     * @param ticks the ticks
     */
    public void lockComponent(int ticks) {
        componentLock.lock(ticks);
    }

    /**
     * Unlock component.
     */
    public void unlockComponent() {
        componentLock.unlock();
    }

    /**
     * Is component locked boolean.
     *
     * @return the boolean
     */
    public boolean isComponentLocked() {
        return componentLock.isLocked();
    }

    /**
     * Lock interactions.
     *
     * @param ticks the ticks
     */
    public void lockInteractions(int ticks) {
        interactionLock.lock(ticks);
    }

    /**
     * Unlock interaction.
     */
    public void unlockInteraction() {
        interactionLock.unlock();
    }

    /**
     * Is interaction locked boolean.
     *
     * @return the boolean
     */
    public boolean isInteractionLocked() {
        return interactionLock.isLocked();
    }

    /**
     * Lock lock.
     *
     * @param key   the key
     * @param ticks the ticks
     * @return the lock
     */
    public Lock lock(String key, int ticks) {
        return lock(key, ticks, null);
    }

    /**
     * Lock lock.
     *
     * @param key    the key
     * @param ticks  the ticks
     * @param elapse the elapse
     * @return the lock
     */
    public Lock lock(String key, int ticks, LockElapse elapse) {
        Lock lock = customLocks.get(key);
        if (lock == null) {
            customLocks.put(key, lock = new Lock());
        }
        lock.setElapse(elapse).lock(ticks);
        return lock;
    }

    /**
     * Unlock.
     *
     * @param key  the key
     * @param node the node
     */
    public void unlock(String key, Node node) {
        unlock(key, true, node);
    }

    /**
     * Unlock.
     *
     * @param key         the key
     * @param cacheRemove the cache remove
     * @param node        the node
     */
    public void unlock(String key, boolean cacheRemove, Node node) {
        Lock lock = customLocks.get(key);
        if (lock != null) {
            lock.unlock();
            if (lock.getElapseEvent() != null) {
                lock.getElapseEvent().elapse(node, lock);
            }
            if (cacheRemove) {
                customLocks.remove(key);
            }
        }
    }

    /**
     * Is locked boolean.
     *
     * @param key the key
     * @return the boolean
     */
    public boolean isLocked(String key) {
        Lock lock = customLocks.get(key);
        return lock != null && lock.isLocked();
    }

    /**
     * Gets equipment lock.
     *
     * @return the equipment lock
     */
    public Lock getEquipmentLock() {
        return equipmentLock;
    }

    /**
     * Sets equipment lock.
     *
     * @param equipmentLock the equipment lock
     */
    public void setEquipmentLock(Lock equipmentLock) {
        this.equipmentLock = equipmentLock;
    }

}

File: Server/src/main/core/game/node/entity/lock/ActionLocks.kt
package core.game.node.entity.lock

import core.game.node.Node
import core.game.world.GameWorld

/**
 * Manages various action-based locks for a game node.
 */
class ActionLocks {

    private val movementLock = Lock()
    private val teleportLock = Lock()
    private val componentLock = Lock()
    private val interactionLock = Lock()
    var equipmentLock: Lock? = null
    private val customLocks: MutableMap<String, Lock> = HashMap()

    /**
     * Locks movement and interactions indefinitely.
     */
    fun lock() {
        lock(Int.MAX_VALUE - GameWorld.ticks)
    }

    /**
     * Locks movement and interactions for a given number of ticks.
     *
     * @param ticks Number of ticks to lock for.
     */
    fun lock(ticks: Int) {
        lockMovement(ticks)
        lockInteractions(ticks)
    }

    /**
     * Unlocks movement and interactions.
     */
    fun unlock() {
        unlockMovement()
        unlockInteraction()
    }

    /**
     * Locks movement for the given number of ticks.
     *
     * @param ticks Number of ticks to lock movement for.
     */
    fun lockMovement(ticks: Int) {
        movementLock.lock(ticks)
    }

    /**
     * Unlocks movement.
     */
    fun unlockMovement() {
        movementLock.unlock()
    }

    /**
     * @return `true` if movement is currently locked.
     */
    fun isMovementLocked(): Boolean = movementLock.isLocked()

    /**
     * Locks teleportation for the given number of ticks.
     *
     * @param ticks Number of ticks to lock teleportation for.
     */
    fun lockTeleport(ticks: Int) {
        teleportLock.lock(ticks)
    }

    /**
     * Unlocks teleportation.
     */
    fun unlockTeleport() {
        teleportLock.unlock()
    }

    /**
     * @return `true` if teleportation is currently locked.
     */
    fun isTeleportLocked(): Boolean = teleportLock.isLocked()

    /**
     * Locks component interactions for the given number of ticks.
     *
     * @param ticks Number of ticks to lock component interaction for.
     */
    fun lockComponent(ticks: Int) {
        componentLock.lock(ticks)
    }

    /**
     * Unlocks component interaction.
     */
    fun unlockComponent() {
        componentLock.unlock()
    }

    /**
     * @return `true` if components are currently locked.
     */
    fun isComponentLocked(): Boolean = componentLock.isLocked()

    /**
     * Locks general interactions for the given number of ticks.
     *
     * @param ticks Number of ticks to lock interaction for.
     */
    fun lockInteractions(ticks: Int) {
        interactionLock.lock(ticks)
    }

    /**
     * Unlocks general interactions.
     */
    fun unlockInteraction() {
        interactionLock.unlock()
    }

    /**
     * @return `true` if general interactions are currently locked.
     */
    fun isInteractionLocked(): Boolean = interactionLock.isLocked()

    /**
     * Locks a custom key-based lock for a given number of ticks.
     *
     * @param key Identifier for the custom lock.
     * @param ticks Number of ticks to lock for.
     * @return The [Lock] instance.
     */
    fun lock(key: String, ticks: Int): Lock = lock(key, ticks, null)

    /**
     * Locks a custom key-based lock for a given number of ticks and sets an optional elapse callback.
     *
     * @param key Identifier for the custom lock.
     * @param ticks Number of ticks to lock for.
     * @param elapse Optional callback when the lock elapses.
     * @return The [Lock] instance.
     */
    fun lock(key: String, ticks: Int, elapse: LockElapse?): Lock {
        val lock = customLocks.getOrPut(key) { Lock() }
        lock.setElapse(elapse).lock(ticks)
        return lock
    }

    /**
     * Unlocks a custom lock and invokes its elapse event if set.
     *
     * @param key Identifier for the custom lock.
     * @param node The node associated with the lock.
     */
    fun unlock(key: String, node: Node) {
        unlock(key, true, node)
    }

    /**
     * Unlocks a custom lock, optionally removing it from cache and invoking its elapse event if set.
     *
     * @param key Identifier for the custom lock.
     * @param cacheRemove Whether to remove the lock from the map after unlocking.
     * @param node The node associated with the lock.
     */
    fun unlock(key: String, cacheRemove: Boolean, node: Node) {
        val lock = customLocks[key] ?: return
        lock.unlock()
        lock.getElapseEvent()?.elapse(node, lock)
        if (cacheRemove) {
            customLocks.remove(key)
        }
    }

    /**
     * Checks if a custom lock is currently active.
     *
     * @param key Identifier for the custom lock.
     * @return `true` if the custom lock is active, otherwise `false`.
     */
    fun isLocked(key: String): Boolean {
        return customLocks[key]?.isLocked() ?: false
    }

}
